sklearn/cluster/tests/test_k_means.py

```diff
@@ -867,3 +867,20 @@ def test_sparse_validate_centers():
     msg = "The shape of the initial centers \(\(4L?, 4L?\)\) " \
           "does not match the number of clusters 3"
     assert_raises_regex(ValueError, msg, classifier.fit, X)
+
+
+def test_less_centers_than_unique_points():
+    X = np.asarray([[0, 0],
+                    [0, 1],
+                    [1, 0],
+                    [1, 0]])  # last point is duplicated
+
+    km = KMeans(n_clusters=4).fit(X)
+
+    # only three distinct points, so only three clusters
+    # can have points assigned to them
+    assert_equal(set(km.labels_), set(range(3)))
+
+    # k_means should warn that fewer labels than cluster
+    # centers have been used
+    assert_warns(ConvergenceWarning, k_means, X, n_clusters=4)
```

Comment:
Can you use assert_warns_message and make sure that it does mention that we find 3 clusters instead 4.